### ISSUE TITLE
Caddy is not matching hostname when calling with IP + Host header (ERROR: no certificate available for IP)

### ISSUE THREAD

**Issue by Chuckame** [caddyserver/caddy#6015]
## Needs
I need to access my home services behind a CGNAT. To solve that, I have the following:

```mermaid
flowchart TD
  aa(remote user) -.->|https://sub.domain.tld| a
  bb(home user) -.->|https://sub.domain.tld| d
subgraph VPS
  a(Caddy server) -->|http| b(tunnel server)
end
  b ===|http tunnel| c(tunnel client)
subgraph home server behind NAT
  c -->|http| d(Caddy server)
  d -->|http| e(app)
end
```

## Problem
When the remote user requests `https://sub.domain.tld` to the `VPS`, then I have a **blank screen** (e.g. no match) 🏳️ 

In parallel, when the home user requests `https://sub.domain.tld` directly to the `home server` (thanks to custom home dns), then I have the web app displayed ✅ 

If I replace the `reverse_proxy` entry `sub.domain.tld` by `:80`, then the remote user is able to reach it 🤯 

It feels like there is an issue with some domain name matcher.

## Reproduction
I'm able to reproduce with just 2 caddy instances on the same host and same network:
```mermaid
flowchart LR
  user -.->|https://sub.domain.tld\nblank answer| a
  user -.->|https://sub.domain.tld\nworking| b
  a(Caddy server A) -->|http| b(Caddy server B)
  b -->|http| c(app)
```

The app used is `mendhak/http-https-echo` to retrieve which headers have been passed to the app (especially for `X-Forwarded-<..>`). After, I also added the `:80` matcher as a fallback and then the app responded well when calling `Caddy server A`. I noticed that the only difference between calling server A (fallbacking on server B through `:80`) or server B (through `http://sub.domain.tld`) is only `X-Forward-Proto` that is respectively `http` and `https`. I added `headers_up X-Forward-Proto https` to server B that fixes the difference, but still going through `:80`.

Here the caddy configs:
```
# Caddy server A
http://sub.domain.tld {
        reverse_proxy http://<ip of Caddy server B>:80
}

# Caddy server B
{
        servers {
              trusted_proxies static <server A network ip>/24
        }
}
http://sub.domain.tld {
        reverse_proxy http://<ip of app>:<port of app>
}
:80 {
        reverse_proxy http://<ip of app>:<port of app>
}
```

_EDIT: added trusted_proxies_

**Comment by francislavoie** [comment]
Add the `debug` global option on both servers and show your logs, make requests with `curl -v` and show that output.

**Comment by Chuckame** [comment]
Oh, I think I have another issue, I'm unable to reach the service with http, even when disabling redirects (and it's still going through the `:80` catchall instead of the `<hostname>` matcher:
```
$ curl -v -H 'Host: <hostname>' http://<hostname> --resolve <hostname>:80:192.168.1.24
* Added <hostname>:80:192.168.1.24 to DNS cache
* Hostname <hostname> was found in DNS cache
*   Trying 192.168.1.24:80...
* Connected to <hostname> (192.168.1.24) port 80 (#0)
> GET / HTTP/1.1
> Host: <hostname>
> User-Agent: curl/8.1.2
> Accept: */*
> 
< HTTP/1.1 503 Service Unavailable
< Server: Caddy
< Date: Thu, 04 Jan 2024 21:52:32 GMT
< Content-Length: 0
< 
* Connection #0 to host <hostname> left intact
```

Here is the access log
```
2024/01/04 21:48:26.909	INFO	http.log.access.log0	handled request	{"request": {"remote_ip": "172.18.0.7", "remote_port": "50668", "client_ip": "172.27.0.1", "proto": "HTTP/1.1", "method": "GET", "host": "<hostname>", "uri": "/api/websocket", "headers": {"Cache-Control": ["no-cache"], "Pragma": ["no-cache"], "Sec-Websocket-Version": ["13"], "Sec-Websocket-Key": ["1zfhKk5H1LI26jONcrsg6w=="], "Upgrade": ["websocket"], "Connection": ["Upgrade"], "Origin": ["https://<hostname>"], "Sec-Websocket-Extensions": ["permessage-deflate; client_max_window_bits"], "X-Forwarded-For": ["172.27.0.1"], "User-Agent": ["Mozilla/5.0 (Linux; Android 14; Pixel 6a Build/UQ1A.231205.015; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/120.0.6099.144 Mobile Safari/537.36 Home Assistant/2023.12.4-11898 (Android 14; Pixel 6a)"], "Accept-Encoding": ["gzip, deflate, br"], "Accept-Language": ["fr-FR,fr;q=0.9,en-US;q=0.8,en;q=0.7"], "X-Forwarded-Host": ["<hostname>"], "X-Forwarded-Proto": ["https"]}}, "bytes_read": 0, "user_id": "", "duration": 0.000127999, "size": 0, "status": 308, "resp_headers": {"Connection": ["close"], "Location": ["https://<hostname>/api/websocket"], "Content-Type": [], "Server": ["Caddy"]}}

```

I enabled debug but there is a lot lot of logs, saying that certs are generated or from cache, the config is valid and loaded. Nothing else.

**Comment by francislavoie** [comment]
> I enabled debug but there is a lot lot of logs

Please share those logs. That's what's important here.

**Comment by Chuckame** [comment]
Here it is
[_caddy-caddy-1_logs.txt](https://github.com/caddyserver/caddy/files/13841952/_caddy-caddy-1_logs.txt)

**Comment by francislavoie** [comment]
That's only the logs from a config reload. We need to see the logs from when you try to make a request.

**Comment by Chuckame** [comment]
Ok, I had a big issue that broke everything.

Now I have this static config:
```
{
  admin off
  servers {
    trusted_proxies static 172.27.0.0/16 172.19.0.0/16
  }

  acme_dns ovh {
    endpoint {env.OVH_ENDPOINT}
    application_key {env.OVH_APP_KEY}
    application_secret {env.OVH_APP_SECRET}
    consumer_key {env.OVH_CONSUMER_KEY}
  }
}

ping.nas.mydomain.tld {
  log {
    format console
  }
  respond "pong"
}
```

And here are the logs:
[_caddy-caddy-1_logs.txt](https://github.com/caddyserver/caddy/files/13845263/_caddy-caddy-1_logs.txt)

Nothing special, except the following one:
```
{"level":"debug","ts":1704478344.0131958,"logger":"tls.handshake","msg":"no matching certificates and no custom selection logic","identifier":"172.19.0.3"}
```
It shows that `172.19.0.3` don't have matching cert, maybe because of the reverse_proxy directive on the predecessor caddy instance:
```
# in docker compose:
   labels:
      caddy_3: ping.nas.chuckame.fr
      caddy_3.reverse_proxy: '{{upstreams https 83}}'

# in caddy instance
ping.nas.chuckame.fr {
	reverse_proxy https://172.27.0.4:83 # ip of the tunnel server
}
```
Then, `172.27.0.4:83` (tunnel server) is forwarded to `172.19.0.2` (tunnel client), forwarded to the home caddy instance `caddy:443` (where ip is `172.19.0.3`)

I also tried with adding the Host header manually, still not working : 
```
ping.nas.domain.tld {
	reverse_proxy https://172.27.0.4:83 {
		header_up Host ping.nas.domain.tld
	}
}
```

**Comment by Chuckame** [comment]
Is there a way of doing the same as curl to use a host name but passing the IP at the same time ? Like this:
```
# curl cmd
curl -v -H 'Host: <hostname>' http://<hostname> --resolve <hostname>:80:<real IP>

# Caddyfile
ping.nas.domain.tld {
	reverse_proxy https://<hostname>:83 {
                 # something like this
                 resolve <hostname> to <IP>
	}
}

**Comment by mohammed90** [comment]
What HTTP tunnel software are you using? For completeness

**Comment by Chuckame** [comment]
I used frp, but now using rathole, both with the same issue. It seems like the home caddy ignoring the host header.

**Comment by tullo** [comment]
Just a thought - what about approaching your end goal using e.g. inlets.dev or ngrok?

**Comment by mohammed90** [comment]
> It seems like the home caddy is ignoring the host header.

That's not accurate. Caddy respects the `Host` header. I'm suspecting the tunnel you're using is manipulating the header in some way.

**Comment by Chuckame** [comment]
Frp were maybe modifying the request. Rathole is just a tunnel with port forwarding, just whatever-data over tcp, that's also why I moved to rathole. 

Just my previous message: https://github.com/caddyserver/caddy/issues/6015#issuecomment-1877816249

I can reproduce the behavior:

**Comment by Chuckame** [comment]
@tullo 
> what about approaching your end goal using e.g. inlets.dev or ngrok?
Because I prefer to host everything myself (as usual for geeks: the challenge & privacy).

I'm also wanting to pass through caddy instead of directly reverse-tunneling the app itself to centralise configuration, auth, rate limits, etc..

@mohammed90 & @francislavoie , what are your ideas after my previous message ? I showed that the domain matching is not done on the host header, but the hostname of the url.

Btw, I've tried the following: Having different hostname and host header, that is not working (going through my `:443` catchall as expected, corresponding to `Perdu! (https)` response):
```
$ curl https://ping.nas.chuckame.fr --resolve ping.nas.mydomain.tld:443:192.168.1.36 -H 'Host: yoloooo.mydomain.tld'
Perdu! (https)
```

**Comment by Chuckame** [comment]
So to sum up:
- only hostname ✅ (`curl https://aaa.mydomain.tld --resolve aaa.mydomain.tld:443:192.168.1.36)
- hostname == host header ✅ (`curl https://aaa.mydomain.tld --resolve aaa.mydomain.tld:443:192.168.1.36 -H 'Host: aaa.mydomain.tld'`)
- only host header ❌ `curl https://192.168.1.36 -H 'Host: aaa.mydomain.tld'`
  - No cert for `192.168.1.36`
- hostname != host header ❌ (`curl https://aaa.mydomain.tld --resolve aaa.mydomain.tld:443:192.168.1.36 -H 'Host: bbb.mydomain.tld'`)
  - Going through the `:443` catchall, because what to do, respond to `aaa` or `bbb` ? But yes, we don't really care, it's just to test all the possibilities

**Comment by mohammed90** [comment]
> @tullo 
> > what about approaching your end goal using e.g. inlets.dev or ngrok?
> Because I prefer to host everything myself (as usual for geeks: the challenge & privacy).
> 
> I'm also wanting to pass through caddy instead of directly reverse-tunneling the app itself to centralise configuration, auth, rate limits, etc..
> 
> @mohammed90 & @francislavoie , what are your ideas after my previous message ? I showed that the domain matching is not done on the host header, but the hostname of the url.
> 
> Btw, I've tried the following: Having different hostname and host header, that is not working (going through my `:443` catchall as expected, corresponding to `Perdu! (https)` response):
> ```
> $ curl https://ping.nas.chuckame.fr --resolve ping.nas.mydomain.tld:443:192.168.1.36 -H 'Host: yoloooo.mydomain.tld'
> Perdu! (https)
> ```
>  

In HTTPS connections, Caddy relies on the [SNI](https://www.cloudflare.com/learning/ssl/what-is-sni/) to find the respective certificate because the [HTTP headers are encrypted](https://stackoverflow.com/a/187679). What you're seeing with curl is because curl sets the SNI to the hostname and not whatever you define in the `Host` header.

I'm still suspecting something awry with rathole, and there are reports of it misbehaving with HAproxy. Granted, I haven't dug deep yet into the reported issues, especially how unfamiliar we are with how rathole operates on the network/protocol level. Research of how rathole operates and how it could affect Caddy is necessary. Perhaps more explanation and logs could be helpful.

**Comment by Chuckame** [comment]
@mohammed90 Ok, so I'll look into it to maybe find another tunnel that respects SNI. After a quick search, the `--resolve` curl param plays on the SNI.

But I still have the issue when using http (no https) by adding the global option `auto_https disable_redirects`, it still goes through the `:80` catchall instead of the hostname:
```
# Caddyfile
{
  admin off
  auto_https disable_redirects
  acme_dns ovh {
    endpoint {env.OVH_ENDPOINT}
    application_key {env.OVH_APP_KEY}
    application_secret {env.OVH_APP_SECRET}
    consumer_key {env.OVH_CONSUMER_KEY}
  }
}
ping.nas.mydomain.tld {
  log {
    format console
  }
  respond "pong"
}
:80 {
  log {
    format console
  }
  respond "Perdu!"
}

$ curl http://192.168.1.36 -H 'Host: ping.nas.mydomain.tld'
Perdu!

```

**Comment by francislavoie** [comment]
`auto_https disable_redirects` doesn't turn off HTTPS, it only removes the HTTP->HTTPS redirect when you make HTTP requests. Caddy is still listening for HTTPS requests with `ping.nas.mydomain.tld`. If you want to handle HTTP requests (not HTTPS) then prefix the site address with `http://` (such as `http://ping.nas.mydomain.tld`).

**Comment by Chuckame** [comment]
> auto_https disable_redirects doesn't turn off HTTPS, it only removes the HTTP->HTTPS redirect when you make HTTP requests. Caddy is still listening for HTTPS requests with ping.nas.mydomain.tld. If you want to handle HTTP requests (not HTTPS) then prefix the site address with http:// (such as http://ping.nas.mydomain.tld).
See my needs section in the original post. I need https locally and access to my service remotely. Then it comes http or https. since https is not working and complexifies a bit things (only SNI), that's why I tried http. But in fact, it doesn't matter if I have http or https between caddy instances 😄

**Comment by Chuckame** [comment]
I tried again with frp, and I'm getting the exact same error on the home caddy instance: `TLS handshake error from 172.19.0.3:48122: no certificate available for '172.19.0.6'`. It feels like the SNI is loosen somewhere.... I opened a ticket on the frp side https://github.com/fatedier/frp/issues/3915

If you have another idea to have `remote https caddy -> some tunneling -> home https caddy` whatever the tool, I would love it

**Comment by mohammed90** [comment]
> It feels like the SNI is loosen somewhere.... I opened a ticket on the frp side fatedier/frp#3915

I found this in frp code after doing a quick search, which seems relevant:

https://github.com/fatedier/frp/blob/dev/pkg%2Fconfig%2Fv1%2Fcommon.go#L94-L96

I'll come back tomorrow for the other bit.

**Comment by Chuckame** [comment]
IT'S WORKING.

Solution:
Replace on remote caddy side from the ip of the tunnel server to the real domain name, and add a network alias on the tunnel server container.

Final config:
remote caddy:
```
sub.mydomain.tld {
	reverse_proxy https://ping.nas.chuckame.fr:84
}
```

remote docker-compose:
```yaml
tunnel-server:
    networks:
      caddy:
        aliases:
          - sub.mydomain.tld
    [...]
```

That's all!

I'll cleanup tomorrow and make a final word, good night!

**Comment by mohammed90** [comment]
I believe this is resolved now and is more of a usage question than a bug in Caddy. I'll close this issue but feel free to elaborate on the solution if you'd like.

In the future, please ask usage questions in the [forum](https://caddy.community).

**Comment by francislavoie** [comment]
You might be looking for `tls_server_name` to override SNI https://caddyserver.com/docs/caddyfile/directives/reverse_proxy#the-http-transport

**Comment by Chuckame** [comment]
tls_server_name seems to really be the final need. It would be great to add an example in the `Reverse proxy to an HTTPS upstream:` section, warning the users that when they forward the https traffic, they also would like to set a custom sni

**Comment by Chuckame** [comment]
Also thinking, what not "forwarding" the upstream sni when the reverse_proxy is about an ip address?
